domain/network/state/subnet_test.go

```diff
@@ -0,0 +1,420 @@
+	err = st.AddSpace(ctx.Background(), spUUID, "space0", "foo", []string{})
+	c.Assert(err, jc.ErrorIsNil)
+
+	subnetUUID0, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID0,
+		"10.0.0.0/24",
+		"provider-id",
+		"provider-network-id-0",
+		0,
+		[]string{"az0", "az1"},
+		spUUID.String(),
+		nil,
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	subnetUUID1, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID1,
+		"10.0.1.0/24",
+		"provider-id",
+		"provider-network-id-1",
+		0,
+		[]string{"az0", "az1"},
+		spUUID.String(),
+		nil,
+	)
+	c.Assert(err, gc.ErrorMatches, "UNIQUE constraint failed: provider_subnet.provider_id")
+}
+
+func (s *stateSuite) TestRetrieveSubnetByUUID(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+
+	// Add a subnet of type base.
+	subnetUUID0, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID0,
+		"192.168.0.0/20",
+		"provider-id-0",
+		"provider-network-id-0",
+		0,
+		[]string{"az0"},
+		"",
+		nil,
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	// Add a subnet of type fan.
+	subnetUUID1, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID1,
+		"10.0.0.0/12",
+		"provider-id-1",
+		"provider-network-id-1",
+		0,
+		[]string{"az1"},
+		"",
+		&network.FanCIDRs{
+			FanLocalUnderlay: "192.168.0.0/20",
+			FanOverlay:       "10.0.0.0/8",
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	// Add a space with subnet base.
+	spUUID, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSpace(ctx.Background(), spUUID, "space0", "provider-space-id", []string{subnetUUID0.String()})
+	c.Assert(err, jc.ErrorIsNil)
+
+	expected := &network.SubnetInfo{
+		ID:                network.Id(subnetUUID0.String()),
+		CIDR:              "192.168.0.0/20",
+		ProviderId:        "provider-id-0",
+		ProviderSpaceId:   "provider-space-id",
+		ProviderNetworkId: "provider-network-id-0",
+		VLANTag:           0,
+		AvailabilityZones: []string{"az0"},
+		SpaceID:           spUUID.String(),
+		SpaceName:         "space0",
+	}
+	sn0, err := st.GetSubnet(ctx.Background(), subnetUUID0)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(sn0, gc.DeepEquals, expected)
+
+	sn1, err := st.GetSubnet(ctx.Background(), subnetUUID1)
+	c.Assert(err, jc.ErrorIsNil)
+	expected = &network.SubnetInfo{
+		ID:                network.Id(subnetUUID1.String()),
+		CIDR:              "10.0.0.0/12",
+		ProviderId:        "provider-id-1",
+		ProviderSpaceId:   "provider-space-id",
+		ProviderNetworkId: "provider-network-id-1",
+		VLANTag:           0,
+		AvailabilityZones: []string{"az1"},
+		SpaceID:           spUUID.String(),
+		SpaceName:         "space0",
+		FanInfo: &network.FanCIDRs{
+			FanLocalUnderlay: "192.168.0.0/20",
+		},
+	}
+	c.Check(sn1, gc.DeepEquals, expected)
+}
+
+func (s *stateSuite) TestRetrieveAllSubnets(c *gc.C) {
```

Comment:
This doesn't test correct Fan population. See comment regarding types.

Reply:
As discussed offline, we don't mind populating `FanInfo` at retrieve time.